feat: adding generic ERC721 withdraw contract

Nilay27 opened the pull request:
* Opening this PR instead of #252 due to the decision of separating each of the withdrawal functions into separate contracts.
* Adding files for shared withdraw erc721 contract.
* Also added ERC721Mintable.sol in the shared folder of packages so that it can be mocked and used by other future package contracts.
* Tests for batchWithdrawERC721 are similar to our usual withdrawERC721 with only adding cases of array length mismatch and withdrawal of multiple ERC721 tokens.
* Also resolved @ramenforbreakfast comments on #252

XavierEkkel commented on packages/prepo-shared-contracts/contracts/WithdrawERC721.sol:
```diff
@@ -0,0 +1,32 @@
+    }
+  }
+
+  function onERC721Received(
```
I don't think we need this - this implies that any contract using this contract will want to accept ERC721s, whereas they may actually just want to withdraw ERC721s that were accidentally sent.

XavierEkkel commented on packages/prepo-shared-contracts/contracts/interfaces/IWithdrawERC721.sol:
```diff
@@ -0,0 +1,7 @@
+
+// TODO: add natspec comments
+interface IWithdrawERC721 {
+  function withdrawERC721(address[] calldata erc721Tokens, uint256[] calldata amounts) external;
```
```suggestion
  function withdrawERC721(address[] calldata erc721Tokens, uint256[] calldata ids) external;
```

XavierEkkel commented on packages/prepo-shared-contracts/contracts/WithdrawERC721.sol:
```diff
@@ -0,0 +1,32 @@
+  {
+    require(_erc721Tokens.length == _ids.length, "Array length mismatch");
+    for (uint256 i; i < _erc721Tokens.length; ++i) {
+      IERC721(_erc721Tokens[i]).safeTransferFrom(address(this), owner(), _ids[i]);
```
`safeTransferFrom` adds unnecessary restriction to the withdrawal, as it requires the `to` address i.e. the owner address (if it's a smart contract) to implement `onERC721Received`, which probably won't be the case for e.g. a Gnosis Safe
let's just make it `transferFrom`

XavierEkkel replied:
for consistency with https://github.com/prepo-io/prepo-monorepo/pull/267#discussion_r935803174, let's add a TODO to add a `_recipients` array

XavierEkkel commented on packages/prepo-shared-contracts/test/WithdrawERC721.test.ts:
```diff
@@ -0,0 +1,186 @@
+      expect(await firstMockERC721.ownerOf(ERC721IdArray[0])).to.not.eq(withdrawERC721.address)
+    })
+
+    it('withdraws different ids of multiple ERC721 tokens', async () => {
```
```suggestion
    it('withdraws different ids of the same ERC721 contract', async () => {
```

XavierEkkel commented on packages/prepo-shared-contracts/test/WithdrawERC721.test.ts:
```diff
@@ -0,0 +1,186 @@
+      expect(await firstMockERC721.ownerOf(ERC721IdArray[1])).to.not.eq(withdrawERC721.address)
+    })
+
+    it('withdraws different ids of multiple ERC721 tokens', async () => {
```
```suggestion
    it('withdraws different ids of multiple ERC721 contracts', async () => {
```

XavierEkkel commented on packages/prepo-shared-contracts/test/WithdrawERC721.test.ts:
```diff
@@ -0,0 +1,186 @@
+    })
+  })
+
+  describe('# onERC721Received', () => {
```
can delete

XavierEkkel reviewed (approved):
LGTM after comments

ramenforbreakfast commented on packages/prepo-shared-contracts/test/WithdrawERC721.test.ts:
```diff
@@ -0,0 +1,186 @@
+      ).revertedWith('Array length mismatch')
+    })
+
+    it('reverts if token id not owned by contract', async () => {
```
Should be consistent with other test suites and declare within this test the not owned tokenId, that way you are not relying on passing the literal value everywhere.

So `const idNotOwnedByContract = 2` within this test, since its the only one that needs it, and then just reference that.

ramenforbreakfast reviewed (approved):
LGTM, besides Xavier's comments and my own nit.

Nilay27 commented:
* As per the suggestions, removed `onERC721Received` and corresponding tests.
* Also changed `safeTransferFrom` to `transferFrom` and included the suggestion to have `_recipient` in the argument for `withdrawERC721`.
